packages/terminal/src/browser/terminal-quick-open-service.ts

```diff
@@ -0,0 +1,160 @@
+} from '@theia/core/lib/browser';
+import { CommandContribution } from '@theia/core/lib/common';
+import { CommandRegistry } from '@theia/core/lib/common';
+import { WidgetManager } from '@theia/core/lib/browser';
+import { TerminalWidget } from './base/terminal-widget';
+import { TerminalService } from './base/terminal-service';
+import { TerminalFrontendContribution, TerminalCommands } from './terminal-frontend-contribution';
+
+@injectable()
+export class TerminalQuickOpenService implements QuickOpenModel, QuickOpenHandler {
+
+    @inject(PrefixQuickOpenService)
+    protected readonly prefixQuickOpenService: PrefixQuickOpenService;
+
+    @inject(WidgetManager)
+    protected readonly widgetManager: WidgetManager;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    // Todo: should be removed once it is refactored.
+    @inject(TerminalFrontendContribution)
+    protected readonly terminalFrontendContribution: TerminalFrontendContribution;
+
+    readonly prefix: string = 'term ';
+
+    get description(): string {
+        return 'Show All Opened Terminals';
+    }
+
+    getModel(): QuickOpenModel {
+        return this;
+    }
+
+    getOptions(): QuickOpenOptions {
+        return {
+            fuzzyMatchLabel: {
+                enableSeparateSubstringMatching: true
+            },
+            fuzzyMatchDescription: {
+                enableSeparateSubstringMatching: true
+            }
+        };
+    }
+
+    open(): void {
+        this.prefixQuickOpenService.open(this.prefix);
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        // Get the list of currently opened terminal widgets, sorted by the widget id.
+        const widgets: TerminalWidget[] = this.terminalService.all
+            .sort((widgetA, widgetB) => (widgetA.id.localeCompare(widgetB.id)));
+
+        for (const widget of widgets) {
+            const item = await this.toItem(widget);
+            terminalItems.push(item);
+        }
+        // Append a quick open item to create a new terminal.
+        const createNewTerminalItem = new QuickOpenItem({
+            label: 'Open New Terminal',
+            run: this.getCreateNewTerminalfunction()
+        });
+        terminalItems.push(createNewTerminalItem);
+
+        acceptor(terminalItems);
+        return;
+    }
+
+    /**
+     * Get the function that can create a new terminal.
+     * @param {TerminalWidget} widget - the terminal widget to be opened.
+     * @returns Function that would create a new terminal if mode === QuickOpenMode.OPEN.
+     */
+    protected getCreateNewTerminalfunction(): (mode: QuickOpenMode) => boolean {
+        return (mode: QuickOpenMode) => {
+            if (mode !== QuickOpenMode.OPEN) {
+                return false;
+            }
+            this.terminalFrontendContribution.openTerminal();
```

Comment:
You can do the following [instead](https://github.com/theia-ide/theia/pull/5577/files#r297383606) and not change the access modifier.